### ISSUE TITLE
Remove internal patches

### ISSUE THREAD

**Issue by johanneswuerbach** [sideroad/testem-multi#10]
What kind of hooks would you need to stop patching testem internals like https://github.com/sideroad/testem-multi/blob/master/lib/coverage_middleware.js#L12?

This patch is broken since `0.6.24` and I can really ensure, that we don't refactor code between releases.

**Comment by piuccio** [comment]
I wrote that code. It works like this:

This [code](https://github.com/sideroad/testem-multi/blob/master/lib/testem-multi.js#L272) adds some custom middlewares to the internal express object.

This can't be done with the proxy configuration because some of them have to be loaded even before testem defaults. The proxy config loads them after, I opened airportyh/testem#341 in the past, but apparently a better solution was coming and that PR was never merged.

The two most important middlewares are
- [coverage.testem](https://github.com/sideroad/testem-multi/blob/master/lib/testem-multi.js#L274) this creates a custom version of testem.js. I'll talk about that later
- [this one](https://github.com/sideroad/testem-multi/blob/master/lib/testem-multi.js#L283) is instead needed to fix airportyh/testem#294

The first one has to be injected before any other middleware and there's no way to do it right now.
The second one instead goes at the end, but errors are swallowed by the last static middleware.

Back to the previous point, I need to create a custom version of testem.js because I need to inject extra JS code in every test page.
Testem multi works both with `configuration` and `custom test page`. When running in custom file (`{"test_page": "tests.html"}`) there's no way to inject in every test an extra script, unless you want to modify every test page.

This [code](https://github.com/sideroad/testem-multi/blob/master/lib/coverage_middleware.js#L22) injects `'../../../../public/testem/istanbul_coverage.js'` into testem.js.
That's why I need to add have middlewares even before testem default ones.
The file listens to `Testem.on("all-test-results")` to send back the coverage to the custom endpoint. The request has to be synchronous or you'll have problems with airportyh/testem#360. I remember having to use a [setTimeout](https://github.com/airportyh/testem/issues/360#issuecomment-68538585) to make it work, but I can't find it in the code, so that's probably working fine with a sync POST.

tl;dr
- inject some code into any test page without asking the user to change the test code
- add error middlewares
- better shut down the server, considering that there might be user code between the end of a test and closing the browser

**Comment by johanneswuerbach** [comment]
Testem has now an official way to add custom middlewares, doesn't this solve both issues? https://github.com/airportyh/testem/pull/410

**Comment by piuccio** [comment]
No, is [this solution](https://github.com/airportyh/testem/pull/410#issuecomment-57691273) was never implemented. Middlewares now go first. I still need to add something after [this one](https://github.com/airportyh/testem/blob/master/lib/server/index.js#L100).
